### PR TITLE
Added count(*) support to sql parser. Fixed warnings in sql parser.

### PR DESCRIPTION
**Pull request by treeform**
* Created new `nkQuotedIdent` node type. 
* Fixed warning related to `toUpperAscii` and `toLowerAscii`.
* Moved stream related methods under when not js block.
* Fix order by limit by parsing.

### CODE REVIEW

File: lib/pure/parsesql.nim [hunk 1]
```diff
@@ -566,18 +573,36 @@ type
...
 
+
 {.deprecated: [EInvalidSql: SqlParseError, PSqlNode: SqlNode,
     TSqlNode: SqlNodeObj, TSqlParser: SqlParser, TSqlNodeKind: SqlNodeKind].}
 
-proc newNode(k: SqlNodeKind): SqlNode =
+proc repr*(node: SqlNode): string =
+  result = $node.kind
+  if node.kind in LiteralNodes:
+    result &= " \"" & node.strVal & "\""
```
Comment (Araq) [inline lib/pure/parsesql.nim]: That is not how to quote things.

Comment (treeform) [reply lib/pure/parsesql.nim]: I don't like how this `repr` works anyways. I'll just remove the whole function for now.

**Comment by treeform** [comment]
I have another problem related to this code I don't know how to fix. I want to use this code in JavaScript (client side SQL validation) but it uses `lexbase` which uses `newStringStream` which does pointer stuff and memory copy and moves so it can't compile. Should I try to make a JS safe lexbase? or JS safe StringStream? I am at a loss. Or try to make this parser not use lexbase and just use plain strings?

**Comment by Araq** [comment]
> Should I try to make a JS safe lexbase?

Yeah, go for it. I'll help you out on IRC.

**Comment by dom96** [comment]
How I solved this in the ``json`` module might be relevant: https://github.com/nim-lang/Nim/blob/devel/lib/pure/json.nim#L1245

File: lib/pure/parsesql.nim [hunk 2]
```diff
@@ -689,9 +702,15 @@ proc parseSelect(p: var SqlParser): SqlNode
 
 proc identOrLiteral(p: var SqlParser): SqlNode =
   case p.tok.kind
-  of tkIdentifier, tkQuotedIdentifier:
+  of tkQuotedIdentifier:
     result = newNode(nkIdent, p.tok.literal)
     getTok(p)
+  of tkIdentifier:
+    if p.tok.literal.toLowerAscii() in SqlFunctions:
```
Comment (Araq) [inline lib/pure/parsesql.nim]: Er, that is not how the SQL is supposed to work. We don't know the SQL functions there can be hundreds per SQL dialect. Why special case these few? It also makes the parser slower, ``tkIdentifier`` is likely on the critical path.

File: lib/pure/parsesql.nim [hunk 3]
```diff
@@ -512,6 +512,7 @@ type
     nkPrefix,
     nkInfix,
     nkCall,
+    nkFunction,
```
Comment (Araq) [inline lib/pure/parsesql.nim]: ``nkFunction`` is a misguided concept.

File: lib/pure/parsesql.nim (hunk 2 above)
Comment (treeform) [reply lib/pure/parsesql.nim]: I removed the functions list, but I had to pass Quoted Identifiers from tokens to nodes as there needs to be a way to say quote `foo as "count"` and also have the unquoted `count(*)`.

File: lib/pure/parsesql.nim (hunk 3 above)
Comment (treeform) [reply lib/pure/parsesql.nim]: removed

**Comment by treeform** [comment]
@dom96 It looks like you just use the native json parser for javascript and then build nim nodes out out it's output?

**Pull request merged**